# Giriş
Merhaba değerli araştırmacılar, bu çalışmamda tensorflow(tf) kütüphanesi üzerinde çalışacağım. Bu çalışmada tensorflow kütüphanesinin [tutorial](https://www.tensorflow.org/tutorials) kaynaklarında faydalanarak hazırlayacağım. Bu çalışmayı yapma nedenim tensorflow kütüphanesi anlamak ve üzerinde araştırmalar yapmak aynı zamanda tensorflow kütühanesi üzerinde olan bilgi darcığımı arttırmak istememdendir. Bu çalışma sürecinde kendi yorumlarımı sizler ile paylaşacağım. Sizlere verdiğim tutorial linkinden de çalışmaları takip edebilir ve oradaki bilgilerlede bilginize bilgi katabilirsiniz.

Başlangıç seviyesi için uygun olacak bu çalışmada tensorflow kütüphanesini kullanarak sınıflandırma işlemi gerçekleştireceğiz. Bu çalışmada model için eğitim ve test verisetinin hazırlanması, tensorflow içerisindeki keras kütüphanesi ile model oluşturma ve oluşturduğumuz model ile eğitim yapmayı göreceksiniz.

Çalışmalarımı Colab üzerinde yapacağımdan dolayı tensorflow kendi bilgisayarıma kurmam gerekmiyor ve colab üzerinde kütüphaneyi kullanacağım. Sizlere önerim tf kütüphanesi anlamayı kolaylaştıracağından dolayı colab üzerinde çalışmalarınıza devam etmenizden yana olacaktır. Daha sonra isterseniz tf kütüphanesini indirir üzerinde çalışmalar yapabilirsiniz. Aynı zamanda colab üzerinde bulunan ekran kartınıda kullanabilirsiniz ama sizlere kısıtlı bir hak tanınacaktır hakkınız dolduğunu ekran kartı üzerinde çalışma yapamazsınız.

Google Colab kullanımı hakkımda yazıma ulaşmak isterseniz [buraya tıklayınız lütfen!](Buraya Link gelecek!)

In [ ]:
import tensorflow as tf
print("Tensorflow verison: ", tf.__version__)

Tensorflow verison:  2.15.0


**Veri setinin yüklenmesi**<br>
Bu aşamada keras içerisindeki **mnist** veri setini yükleyeceğiz. Veri setini eğitim ve test olarak parçalara ayırcağız.<br>
**x_train** : Eğitim verisi<br>
**x_test** : Test verisi<br>
**y_train** : Eğitim verisinin sınıfları<br>
**y_test** : Test verisinin sınıfları

In [ ]:
# Veri setini yükle
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 0s 0us/step


In [ ]:
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


60 bin 28x28 boyutunda resim x_train veri kümesi içerisinde bulunmaktadır.<br>
60 bin adet resimin sınıfları y_train veri seti içerisinde bulunmaktadır.<br>

In [ ]:
# Modelin oluşturulması
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape = (28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation = "softmax")
])

modelimizi oluşturuken tensorflow içerisindeki keras kütüphanesinden faydalandık.<br>
**Flatten** : Sinir ağımızın ilk katmanı bu katmanda gelecek verinin boyutunu belirledik.((28, 28))<br>
**Dense** : Yapay sinir ağı. Dense katmanı gelecek veriyi "relu" aktivasyon foksiyonunu kullanarak süzgeçten geçirecek.<br>
**Dropout** : Unutma katmanı. Bu katmanda öğrenilen verini 0.2'sini unutacak ve diğer adımda öğrenmeye devam edecek. Bu katmanı kullanma sebebimiz aşırı öğrenmenin(ezberlemek) önüne geçmek olacaktır.<br>
Son katmanda **Dense** katmanında **softmax** aktivasyon foksiyonunu kullandım çünkü bu son katman gelen verinin sınıfının tahmin katmanıdır. 10 sınıf sayısıdır.

In [ ]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True)

model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

**loss_fn** : Modelin kayıp fonksiyonudur. Bu katman modelin gerçek değerden ne kadar uzak olduğunu ölçer ve modeli güncellemek için kullanılır.

**model.compile** :  Bu kısımda birden çok parametre bulunmakta kısaca ne olduklarına değineyim.
  - **optimizer** : Modelin günceleyecek fonksiyondur, adaptif moment tahmini (Adam) optimizasyon algoritmasıdır ve genellikle çeşitli derin öğrenme görevlerinde etkili olmuştur.
  - **loss** : kayıp fonksiyonudur.
  - **metrics** : modelin değerlendirmesinde kullanacağım metriktir.

In [ ]:
# Eğitim
# Epoch : eğitim adımıdır. Eğitim kaç adımdan oluşacağını belirleriz.
hist = model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 9s 4ms/step - loss: 2.6134 - accuracy: 0.7462
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5831 - accuracy: 0.8462
Epoch 3/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.4715 - accuracy: 0.8733
Epoch 4/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.4258 - accuracy: 0.8873
Epoch 5/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.4016 - accuracy: 0.8953


In [ ]:
# Modelin değerlendirme aşaması
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - loss: 0.2654 - accuracy: 0.9335 - 631ms/epoch - 2ms/step


[0.2653728127479553, 0.9334999918937683]

**Loss** : 0.26 , **accuracy** : 0.93 skorlarımız gayet güzel çıktı accuracy değerinin giderk artması ve yüksek çıkması model eğitim için iyidir. Loss değerinin girek düşmesi ve düşük çıkması yine modelimizin için iyidir. Bu durumların sağlanması sonucunda modelimizin öğrendiği gözlemleyebiliriz.

# Sonuç
Arkadaşlar bu çalışmada Tensorflow kütüphanesinin tutorial kısmının Quickstart for begginers kısmını bitirdik. Tutorial kısmında bizden isteneleri yerine getirdik ve öğrenme aşamasının ilk adımını tamamladık. Diğer adımlarda görüşmek üzere öğrenmeye ve kendimi geliştirmeye devam!

Aşağıdaki Linklerden beni takip edebilir ve yapacağım çalışmalardan haberdar olabilirsiniz!<br>
[Linkedin](https://www.linkedin.com/in/ihsancenkiz/)<br>
[Github](https://github.com/ihsncnkz)<br>
[Kaggle](https://www.kaggle.com/ihsncnkz)